### DISCUSSION TITLE
Trigger multiple uploads, show file dialog only once

### DISCUSSION THREAD

**Discussion by daniperez** [Waboodoo/HTTP-Shortcuts#270] [Q&A]
Hi!

I was wondering if the following use case can be implemented with HTTP Shortcuts:

1. User executes a shortcut
2. File dialog appears and user selects several files
3. Every file triggers an upload somewhere (e.g. _one_ `POST` _per file_)

I've been able to accomplish something similar with a regular shortcut with multiple files in the request body that then triggers itself or other shortcuts (once per file upload, as per ["Can I send multiple requests with one shortcut?"](https://http-shortcuts.rmy.ch/faq)). But for every file, it always shows again the file dialog. I guess the problem is that HTTP Shortcuts always shows a file dialog when a file is used in the request body, without a chance to disable that under some conditions (which is understandable, for the sake of simplicity).

Any ideas about how to accomplish this? Thanks!

**Comment by Waboodoo** [comment]
So let me summarize, to see if I understand this scenario correctly: you want to make multiple POST requests, either by running multiple shortcuts or by running the same shortcut multiple times, and each of these requests would upload a file. And instead of having each shortcut open its own file picker, you would want to pick multiple files at once using a single file picker. And the number of requests made / shortcuts triggered would dynamically change with the number of files selected initially. Is that correct? Or would you need to use the same file for multiple requests?

This is currently not possible with the app. Unlike with variables, files (as picked with the file picker) can not be persisted and can not be passed from one shortcut to another when triggering/enqueueing. Each shortcut that uses files (either as form parameters or as the request body) will always request files through the picker on-demand.

I see how this might be useful, so I think it would for sure be something that could fit into the app as a new feature. Currently however I don't yet know how this could be achieved without adding too much complexity and in a way that it's easy to discover and understand. I'm open to suggestions.

**Reply by daniperez** [reply]
Your understanding of the question is correct 👍

> And the number of requests made / shortcuts triggered would dynamically change with the number of files selected initially. Is that correct? Or would you need to use the same file for multiple requests?

I meant the former: I would select some files (say, photos in my phone), and I would like to trigger an upload for each file to a photo or backup site, for example.

> I'm open to suggestions.

I don't know the code but I wanted to explain what I attempted, if only to show what a (naive) user like me tried. Hopefully that can give you some information of the devil your users are trying to do :rofl: 

My first attempt, was to define a shortcut using a "multiple files" parameter (form-data) in the request body which would be stored in, say, `{requestSelectedFiles}`. Little detour: at this point I didn't understand what would be the difference from the built-in `selectedFiles` and whether it would make a difference if I chose a "static variable" for that or just provide a name to the parameter . Anyhow, I used a static variable for that and it seemed to work: I could reference that variable within my javascript. End of little detour 🙂 Once I had that variable defined, my naive solution was to call the same shortcut with a different value for that variable in the "Run on success" script:

```javascript
// -- Run on success --

var requestSelectedFiles = getVariable("requestSelectedFiles")

// Drop the first file, already processed in "Run before execution"
requestSelectedFiles = requestSelectedFiles.shift()

// Tried to set the value of the variable...
setVariable("requestSelectedFiles", requestSelectedFiles)

// ... and also to call the same shortcut with a list 
// of files, one element smaller:
triggerShortcut(shortcut.name, {"requestSelectedFiles": requestSelectedFiles})
```

And my _naive_ expectation was that some of that would tell HTTP Shortcuts "hey, no need to ask for the files, that parameter is already defined".

I was also messing with `selectedFiles`, but that didn't work either.

Again, I'm just explaining what a naive user tried. I understand what I am trying is neither easy to accomplish nor simple to implement. Thanks again!

**Reply by Waboodoo** [reply]
Thanks for the detailed explanation. I'll thinking about this and see if I can find a solution for this problem and implement something in a future release. I can't make any promises though regarding timing. I'll post here if I have any updates.

**Reply by Waboodoo** [reply]
There is now a new (somewhat experimental) feature available in the 2.16.0 release: you can forward one or more selected files to another shortcut via the `enqueueShortcut` function. There is still no possibility to open a file picker via Scripting directly, but you can work around that by just adding a files parameter to your shortcut and then using the `abort()` function in your script to stop the shortcut from making any HTTP request itself, and instead you use `enqueueShortcut` to send the selected file to another shortcut. Since the shortcut will never actually make the HTTP request you can use a dummy URL like `https://example.com`.

In a concrete example, your script (inserted in the "Execute before Shortcut" section) could then look something like this:
```js
// Get all the IDs of the files that were selected
const myFileIds = selectedFiles.map(file => file.id);

// Iterate over all the file IDs
myFileIds.forEach(fileId => {
    // For each file ID, enqueue the execution of the 'my file shortcut', and pass the file ID to it via the special '$files' variable.
    enqueueShortcut('my file shortcut', {
        '$files': fileId,
    });
});

// Stop the execution of the current shortcut, because we don't actually need to make any HTTP request
abort();
```

This will mean that the 'my file shortcut' Shortcut will be executed once for each file that you selected, and it will receive the file as input and will use it instead of opening its own file picker, for a file form parameter or for the request body, depending on what you configured for that shortcut.

Find more documentation for this here: https://http-shortcuts.rmy.ch/scripting#files and https://http-shortcuts.rmy.ch/scripting#trigger-shortcut.

I hope this helps with your usecase. Let me know if there's any issues.

**Reply by daniperez** [reply]
Thanks @Waboodoo ! I tried and it works :clap: :clap: :clap: 

I'll post soon the shortcuts, in case it can be useful for someone else.

**Reply by daniperez** [reply]
> I'll post soon the shortcuts, in case it can be useful for someone else.

Here we go!

https://gist.github.com/daniperez/158ae75cd2c660524e4fb7fc911a1a65

It doesn't feel like an off-the-shelf solution, but it can perhaps unlock someone trying to achieve the same.
